_source/_posts/2019-11-19-secure-kafka-java.md

```diff
@@ -0,0 +1,458 @@
+        this.template = template;
+        this.myTopicConsumer = myTopicConsumer;
+    }
+
+    @GetMapping("/kafka/produce")
+    public void produce(@RequestParam String message) {
+        template.send("myTopic", message);
+    }
+
+    @GetMapping("/kafka/messages")
+    public List<String> getMessages() {
+        return myTopicConsumer.getMessages();
+    }
+
+}
+ ```
+
+This class now has a new endpoint to display the messages stored in your consumer. When this endpoint is called, it sends the current messages it already processed from the Kafka topic. 
+
+Your Java app now has both a Kafka producer and a consumer, so let's test it all together! Restart your application, and go to <http://localhost:8080/kafka/messages>.
+
+Right now, no information is being returned. The reason is pretty simple: your consumer is configured only to receive new messages and you haven't sent a new message yet. Let's fix this problem by going to your web browser and accessing <http://localhost:8080/kafka/produce?message=Message sent by my App!>.
+
+When Kafka receives the message, it will let your consumer know about it right away. Go ahead and go to <http://localhost:8080/kafka/messages> in your browser. You will now see that your message was successfully received!
+
+{% img blog/kafka-java/kafka-consumer-messages.png alt:"Kafka Consumer Message" width:"288" %}{: .center-image }
+
+Great job! You have a Java app capable of producing and consuming messages from Kafka! Before we call it a day though, there is one last step and it's a very important one.
+
+## Secure Your Java Kafka Application
+
+Your app is not very secure right now. Although you are prepared to handle many messages in a distributed environment, those messages are still available to anyone who can find the link to your endpoints. This is a critical vulnerability, so let's make sure it's addressed the right way. 
+
+You're going to use OAuth 2.0 to make sure only authenticated users can see your endpoints. The best part? It's going to take only 5 minutes to add this feature in your app by using Okta to authenticate your users!
+
+### Create an Okta Account
+
+If you don't already have an Okta account, [go ahead and create one](https://developer.okta.com/signup). After completing registration, go through the following steps:
+
+* Login into your account
+* Go to  **Applications** > **Add Application**. You will be redirected to the following page:
+
+{% img blog/kafka-java/okta-applications.png alt:"Okta application page" width:"800" %}{: .center-image }
+
+* Select **Web** and click **Next** 
+* Fill in the following options in the form:
+    - Name: `Bootiful Kafka`
+    - Base URIs: `http://localhost:8080`
+    - Login redirect URLs: `http://localhost:8080/login/oauth2/code/okta`
+* Click **Done**
+
+Now that you have your Okta application, you can use it to authenticate users in your Java + Kafka app.
+
+## Secure Your Java App with User Auth
```

Comment:
Should this be: `## Secure Your Java App with OAuth 2.0`?

Reply:
This was Brent's suggestion, so I'm going to leave it.